src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs

```diff
@@ -338,6 +329,38 @@ static SearchValues<string> PickAhoCorasickImplementation<TCaseSensitivity>(AhoC
                 : new SingleStringSearchValuesFallback<SearchValues.FalseConst>(value, uniqueValues);
         }
 
+        private static SearchValues<string>? TryCreateSingleValuesThreeChars<TValueLength>(
+            string value,
+            HashSet<string>? uniqueValues,
+            bool ignoreCase,
+            bool allAscii,
+            bool asciiLettersOnly)
+            where TValueLength : struct, IValueLength
+        {
+            if (!ignoreCase)
+            {
+                return new SingleStringSearchValuesThreeChars<TValueLength, CaseSensitive>(uniqueValues, value);
+            }
+
+            if (asciiLettersOnly)
+            {
+                return new SingleStringSearchValuesThreeChars<TValueLength, CaseInsensitiveAsciiLetters>(uniqueValues, value);
+            }
+
+            if (allAscii)
+            {
+                return new SingleStringSearchValuesThreeChars<TValueLength, CaseInsensitiveAscii>(uniqueValues, value);
+            }
+
+            // When ignoring casing, all anchor chars we search for must be ASCII.
+            if (char.IsAscii(value[0]) && value.AsSpan().LastIndexOfAnyInRange((char)0, (char)127) > 0)
+            {
+                return new SingleStringSearchValuesThreeChars<TValueLength, CaseInsensitiveUnicode>(uniqueValues, value);
```

Comment:
I'm unclear about this logic:
- Why do we only do this if the first character is ASCII?
- Why do we search from the end rather than from the beginning?
- Why don't we always use this CaseInsensitiveUnicode as a fallback?

Reply:
`SingleStringSearchValuesThreeChars` doesn't have logic to handle non-ASCII case conversion, so we require that the anchor characters are Ascii.

Right now we're always selecting the first character as one of the anchors, and we need at least two.
The check above could alternatively be `char.IsAscii(value[0]) && value.AsSpan(1).ContainsAnyInRange((char)0, (char)127)`

Reply:
Ok. Can you add some more comments? Thanks.